integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/columnLineage/FieldDependenciesCollector.java

```diff
@@ -0,0 +1,74 @@
+package io.openlineage.spark3.agent.lifecycle.plan.columnLineage;
+
+import io.openlineage.spark.agent.util.ScalaConversionUtils;
+import io.openlineage.spark3.agent.lifecycle.plan.columnLineage.customCollectors.CustomFieldDependencyCollector;
+import io.openlineage.spark3.agent.lifecycle.plan.columnLineage.customCollectors.IcebergMergeIntoFieldDependencyCollector;
+import io.openlineage.spark3.agent.lifecycle.plan.columnLineage.customCollectors.UnionFieldDependencyCollector;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.spark.sql.catalyst.expressions.ExprId;
+import org.apache.spark.sql.catalyst.expressions.Expression;
+import org.apache.spark.sql.catalyst.expressions.NamedExpression;
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression;
+import org.apache.spark.sql.catalyst.plans.logical.Aggregate;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.catalyst.plans.logical.Project;
+
+/**
+ * Traverses LogicalPlan and collects dependencies between the expressions and operations used
+ * within the plan.
+ */
+@Slf4j
+public class FieldDependenciesCollector {
+
+  private final LogicalPlan plan;
+  private final List<CustomFieldDependencyCollector> collectors =
+      Arrays.asList(
+          new UnionFieldDependencyCollector(), new IcebergMergeIntoFieldDependencyCollector());
```

Comment:
Is there a more graceful way to do this? This pattern keeps popping up and it feels harder to extend. What happens when I want to add a Databricks proprietary DependencyCollector or a DeltaDependencyCollector? It feels like this goes against the Open Close principle.

Reply:
https://github.com/OpenLineage/OpenLineage/issues/738